### PR TITLE
Fix AOT build

### PR DESCRIPTION
**Pull request by grendello**
The `leeroy` make target failed with a "file not found" exception while
attempting to invoke the `cross-arm` crosscompiler and the reason for this was
that `cross-arm` was never built because of the `armeabi-v7a` target missing in
the `ALL_AOT_ABIS` make variable in the `build-tools/scripts/BuildEverything.mk`
file.

This commit adds the missing target architecture and fixes the build.

### CODE REVIEW

**Comment by grendello** [comment]
@radekdoulik yeah, it should... :)

**Comment by grendello** [comment]
Hmm, it seems we don't create a new emulator image because some emulator's already running:

```
Task "Xamarin.Android.Tools.BootstrapTasks.CheckAdbTarget" (TaskId:93)
  Task Parameter:ToolExe=adb (TaskId:93)
  Task Parameter:ToolPath=/Users/builder/android-toolchain/sdk/platform-tools (TaskId:93)
  Task Parameter:Timeout=60000 (TaskId:93)
  /Users/builder/android-toolchain/sdk/platform-tools/adb  shell getprop ro.build.version.sdk  (TaskId:93)
  * daemon not running; starting now at tcp:5037 (TaskId:93)
  * daemon started successfully (TaskId:93)
  28 (TaskId:93)
  /Users/builder/android-toolchain/sdk/platform-tools/adb  shell pm path com.android.shell  (TaskId:93)
  package:/system/priv-app/Shell/Shell.apk (TaskId:93)
  Output Property: _ValidAdbTarget=True (TaskId:93)
Done executing task "Xamarin.Android.Tools.BootstrapTasks.CheckAdbTarget". (TaskId:93)
Task "CreateAndroidEmulator" skipped, due to false condition; ( '$(_ValidAdbTarget)' != 'True' ) was evaluated as ( 'True' != 'True' ).
Task "StartAndroidEmulator" skipped, due to false condition; ( '$(_ValidAdbTarget)' != 'True' ) was evaluated as ( 'True' != 'True' ).
```

**Comment by grendello** [comment]
So the problem here appears to be that the first two test suites create the results file in the external storage which creates the files with permissions that make them world-readable. The last 2 result files are [created in the private data store](https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.NUnitLite/Gui/Instrumentations/TestSuiteInstrumentation.cs#L104-L118) with stricter permissions that make them only owner-readable. Thus we can't access them.

**Comment by grendello** [comment]
After examining the emulator image I noticed that the successful test suites have the following entry in their `AndroidManifest.xml` file:

```xml
<uses-sdk android:minSdkVersion="14" android:targetSdkVersion="22" />
```

and the failing ones have instead

```xml
<uses-sdk android:minSdkVersion="14" />
```

Considering that the code which [constructs the results path](https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.NUnitLite/Gui/Instrumentations/TestSuiteInstrumentation.cs#L104-L118) targets Android API 19 and newer in its crucial part, the theory is that the lack of the target SDK version causes Android to behave in a legacy manner - [to return `null` for this code](https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.NUnitLite/Gui/Instrumentations/TestSuiteInstrumentation.cs#L109) and, thus, create the result file path in the app's private storage. This paired with stricter Android permission handling in newer Androids leads to the issue. That's the theory, let's see if it's confirmed after the current build completes...

**Comment by grendello** [comment]
Addition of `android:targetSdkVersion="22"` didn't help, however the latest update appears to have worked locally. Namely, I added external storage permissions to the test suites which failed previously - this affects the path where the results file is placed. We'll know if it really worked in 7+ hours...

**Comment by grendello** [comment]
So, the APK tests failed again, as before, despite the permission changes. However, it's not because the changes weren't effective - they were, as the tests run locally without a hitch (with a single failure across the entire test suite, not related to this PR). However, it seems the problem lies in that the emulator we start on the build bots is never stopped. In effect, the next PR build sees a running emulator and uses it for the tests, which results in:

```
  Task Adb (TaskId:103)
    Arguments:   install "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.Locale_Tests-Signed.apk" (TaskId:103)
  /Users/builder/android-toolchain/sdk/platform-tools/adb   install "/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.Locale_Tests-Signed.apk"  (TaskId:103)
  adb: failed to install /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.Locale_Tests-Signed.apk: Failure [INSTALL_FAILED_INSUFFICIENT_STORAGE] (TaskId:103)
/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/TestApks.targets(150,5): error MSB6006: "adb" exited with code 1. [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/RunApkTests.targets]
    [Output] Output: (TaskId:103)
      adb: failed to install /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.Locale_Tests-Signed.apk: Failure [INSTALL_FAILED_INSUFFICIENT_STORAGE] (TaskId:103)
```

And then further errors stemming from the one above. We fail to kill the emulator instance after the tests complete:

```
Task "Xamarin.Android.Tools.BootstrapTasks.Adb" skipped, due to false condition; ( '$(_EmuTarget)' != '' ) was evaluated as ( '' != '' ).
Task "KillProcess" skipped, due to false condition; ( '$(_EmuTarget)' != '' ) was evaluated as ( '' != '' ).
Task "Xamarin.Android.Tools.BootstrapTasks.Adb" (TaskId:126)
  Task Parameter:Arguments=kill-server (TaskId:126)
  Task Parameter:ToolExe=adb (TaskId:126)
  Task Parameter:ToolPath=/Users/builder/android-toolchain/sdk/platform-tools (TaskId:126)
  Task Parameter:Timeout=60000 (TaskId:126)
  Task Adb (TaskId:126)
    Arguments: kill-server (TaskId:126)
  /Users/builder/android-toolchain/sdk/platform-tools/adb kill-server  (TaskId:126)
    [Output] Output: (TaskId:126)
Done executing task "Xamarin.Android.Tools.BootstrapTasks.Adb". (TaskId:126)
Task "Exec" skipped, due to false condition; ( '$(HostOS)' != 'Windows' And '$(_EmuTarget)' != '' ) was evaluated as ( 'Darwin' != 'Windows' And '' != '' ).
Task "Sleep" skipped, due to false condition; ( '$(HostOS)' != 'Windows' And '$(_EmuTarget)' != '' ) was evaluated as ( 'Darwin' != 'Windows' And '' != '' ).
Task "Exec" skipped, due to false condition; ( '$(HostOS)' != 'Windows' And '$(_EmuTarget)' != '' ) was evaluated as ( 'Darwin' != 'Windows' And '' != '' ).
Done building target "ReleaseAndroidTarget" in project "RunApkTests.targets".: (TargetId:22)
```

I'm going to see why the `_EmuTarget` property is empty while it shouldn't be...

**Comment by grendello** [comment]
build

**Comment by grendello** [comment]
build

**Comment by grendello** [comment]
Made changes to the Jenkins PR builder to kill all emulator instances and remove all test AVDs before the build, let's wait to see if it makes a difference of any kind

**Comment by grendello** [comment]
build

**Comment by grendello** [comment]
This latest update rewrites the `Adb` task + tasks which derive from it so that the `Adb` task no longer uses `PathToolTask` as its base but runs `adb` on its own. The reason is that we want to be able to ignore adb failure exit codes (I know `ToolTask` has a way to do it but it's not very subtle, I wanted more finely grained control over this) and also the ability to log output of the command to a file - this is to work around the "permission denied" error when pulling test results from the private location (still no clue why it works locally but fails on the PR builders). We do it by running the `cat` command as the user who owns the test package - that gives us the permission to read the file but we need to "redirect" output to a local file - this is implemented by the latest update as well.

**NOTE!!** I wasn't able to test the changes on the runtime as currently my local msbuild is broken and fails to run our tests - thus the `do-not-merge` label.

**Comment by grendello** [comment]
Managed to fix my msbuild and run the tests. Just pushed an update that fixes a number of issues I've found. Locally the code appears to work fine:

```
AcquireAndroidTarget:
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb  shell getprop ro.build.version.sdk
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb  shell pm path com.android.shell
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb  shell setprop debug.mono.log timing
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb  logcat -G 4M
UndeployTestApks:
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   uninstall "Mono.Android_Tests"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   uninstall "Xamarin.Android.Bcl_Tests"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   uninstall "Xamarin.Android.JcwGen_Tests"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   uninstall "Xamarin.Android.Locale_Tests"
DeployTestApks:
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   install "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/Mono.Android_Tests-Signed.apk"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   install "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.Bcl_Tests-Signed.apk"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   install "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.JcwGen_Tests-Signed.apk"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   install "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/Xamarin.Android.Locale_Tests-Signed.apk"
RecordApkSizes:
  stat -c "stat: %s %N" "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/Mono.Android_Tests-Signed.apk" > "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/apk-sizes-Mono.Android_Tests-Debug.txt"
  unzip -l "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/Mono.Android_Tests-Signed.apk" >> "/home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/apk-sizes-Mono.Android_Tests-Debug.txt"
RunTestApks:
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Mono.Android_Tests android.permission.READ_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Mono.Android_Tests android.permission.WRITE_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Bcl_Tests android.permission.READ_PHONE_STATE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Bcl_Tests android.permission.READ_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Bcl_Tests android.permission.WRITE_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.JcwGen_Tests android.permission.READ_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.JcwGen_Tests android.permission.WRITE_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Locale_Tests android.permission.READ_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell pm grant Xamarin.Android.Locale_Tests android.permission.WRITE_EXTERNAL_STORAGE
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell am instrument  -e "loglevel Verbose" -w "Mono.Android_Tests/xamarin.android.runtimetests.TestInstrumentation"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   logcat -v threadtime -d
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell run-as Mono.Android_Tests cat "/storage/emulated/0/Android/data/Mono.Android_Tests/files/Documents/TestResults.xml"
    writing stdout to file: /home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/TestResult-Mono.Android_Tests.xml
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell am instrument  -e "loglevel Verbose" -w "Xamarin.Android.Bcl_Tests/xamarin.android.bcltests.XUnitInstrumentation"
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   logcat -v threadtime -d
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell run-as Xamarin.Android.Bcl_Tests cat "/storage/emulated/0/Android/data/Xamarin.Android.Bcl_Tests/files/Documents/TestResults.xUnit.xml"
    writing stdout to file: /home/grendel/vc/xamarin/xamarin-android/tests/../bin/TestDebug/TestResult-Xamarin.Android.Bcl_Tests.xunit.xml
  Executing: /home/grendel/android-toolchain/sdk/platform-tools/adb   shell am instrument  -e "loglevel Verbose" -w "Xamarin.Android.Bcl_Tests/xamarin.android.bcltests.NUnitInstrumentation"
```

Note the `run-as` invocations - this **should** make it work on the PR builder for the private data location. We'll see :)

File: tests/locales/Xamarin.Android.Locale-Tests/Xamarin.Android.Locale-Tests.csproj [hunk 1]
```diff
@@ -18,12 +18,10 @@
     <AndroidManifest>Properties\AndroidManifest.xml</AndroidManifest>
     <AndroidSupportedAbis>armeabi-v7a;x86</AndroidSupportedAbis>
     <AndroidUseSharedRuntime>False</AndroidUseSharedRuntime>
-    <AndroidUseLatestPlatformSdk>false</AndroidUseLatestPlatformSdk>
+    <TargetFrameworkVersion>v8.1</TargetFrameworkVersion>
+    <AndroidUseLatestPlatformSdk>true</AndroidUseLatestPlatformSdk>
   </PropertyGroup>
```
Comment (radekdoulik) [inline tests/locales/Xamarin.Android.Locale-Tests/Xamarin.Android.Locale-Tests.csproj]: What is the reasoning for this change? Wouldn't it limit the tests, when nearly everything would target v8.1?

Comment (grendello) [reply tests/locales/Xamarin.Android.Locale-Tests/Xamarin.Android.Locale-Tests.csproj]: This is just to bring parity with other tests - note however line 22, it renders line 21 ineffective by choosing whatever the latest platform is. The test projects must be somehow made to use the same project options, but that's a material for a different PR.

**Comment by grendello** [comment]
Timeout while deploying APK to the emulator... will build again :(

**Comment by grendello** [comment]
build

**Comment by grendello** [comment]
The failed status is because of 

```
(RunTestApks target) ->
  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/TestApks.targets(183,5): error :   Command /Users/builder/android-toolchain/sdk/platform-tools/adb -s emulator-5570  shell run-as Mono.Android_Tests cat "/data/user/0/Mono.Android_Tests/files/.__overrid
e__/TestResults.xml" failed with exit code 1 [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/RunApkTests.targets]


"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets" (RunAllTests target) (1) ->
"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets" (RunApkTests target) (1:4) ->
"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/RunApkTests.targets" (default target) (4:2) ->
  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/TestApks.targets(183,5): error :   Command /Users/builder/android-toolchain/sdk/platform-tools/adb   shell run-as Mono.Android_Tests cat "/data/user/0/Mono.Android_Tests/files/.__override__/TestResults.
xml" failed with exit code 1 [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/RunApkTests.targets]


"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets" (RunAllTests target) (1) ->
"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/RunTests.targets" (RunApkTests target) (1:4) ->
"/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/RunApkTests.targets" (default target) (4:3) ->
  /Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/build-tools/scripts/TestApks.targets(183,5): error :   Command /Users/builder/android-toolchain/sdk/platform-tools/adb   shell run-as Mono.Android_Tests cat "/data/user/0/Mono.Android_Tests/files/.__override__/TestResults.
xml" failed with exit code 1 [/Users/builder/jenkins/workspace/xamarin-android-pr-builder/xamarin-android/tests/RunApkTests.targets]

```

But I see no actual error in the logs. Just the exit status. The failed tests are for the `Release` build and they DID succeed (just as their `Debug` run) - copying of the results file was what failed. As such I'd love for this PR to be merged... I can't find what's wrong with the PR builder and all of those tests DO work (and I get no failures copying result files when they run locally, too)

**Pull request merged**